base/special/cbrt.jl

```diff
@@ -0,0 +1,144 @@
+"""
+@inline function _approx_cbrt(x::T) where {T<:Union{Float32,Float64}}
+    adj = exponent_bias(T)*2/3 - 0.03306235651
+    k = significand_bits(T) - (8*sizeof(T) - 32)
+
+    u = highword(x) & 0x7fff_ffff
+    if u >= Base.Math.highword(realmin(T))
+        v = div(u, UInt32(3)) + floor(UInt32, adj * 2.0^k)
+    else
+        # subnormal
+        x *= maxintfloat(T)
+        adj -= exponent(maxintfloat(T))/3
+        u = highword(x) & 0x7fff_ffff
+        v = div(u, UInt32(3)) + floor(UInt32, adj * 2.0^k)
+    end
+    return copysign(fromhighword(T, v), x)
+end
+
+@inline function _improve_cbrt(x::Float32, t::Float32)
+    # Newton iterations solving
+    #   t^2 - x/t == 0
+    # with update
+    #   t <- t*(t^3 + 2*x)/(2*t^3 + x)
+
+    # Use double precision so that its terms can be arranged for efficiency
+    # without causing overflow or underflow.
+    xx = Float64(x)
+    tt = Float64(t)
+
+    # 1st step: 16 bits accuracy
+    tt3 = tt^3
+    tt *= (2*xx + tt3)/(x + 2*tt3)
+
+    # 2nd step: 47 bits accuracy
+    tt3 = tt^3
+    tt *= (2*xx + tt3)/(x + 2*tt3)
+
+    return Float32(tt)
+end
+
+@inline function _improve_cbrt(x::Float64, t::Float64)
+    # cbrt to 23 bits:
+    #
+    #    cbrt(x) = t * cbrt(x / t^3) ~= t * P(t^3 / x)
+    #
+    # where P(r) is a polynomial of degree 4 that approximates 1/cbrt(r)
+    # to within 2^-23.5 when |r - 1| < 1/10.  The rough approximation
+    # has produced t such than |t/cbrt(x) - 1| ~< 1/32, and cubing this
+    # gives us bounds for r = t^3/x.
+
+    r = (t*t)*(t/x)
+    t *= (@horner(r, 1.87595182427177009643, -1.88497979543377169875, 1.621429720105354466140) +
+          r^3 * @horner(r, -0.758397934778766047437, 0.145996192886612446982))
+
+    # Round t away from zero to 23 bits (sloppily except for ensuring that
+    # the result is larger in magnitude than cbrt(x) but not much more than
+    # 2 23-bit ulps larger).  With rounding towards zero, the error bound
+    # would be ~5/6 instead of ~4/6.  With a maximum error of 2 23-bit ulps
+    # in the rounded t, the infinite-precision error in the Newton
+    # approximation barely affects third digit in the final error
+    # 0.667; the error in the rounded t can be up to about 3 23-bit ulps
+    # before the final error is larger than 0.667 ulps.
+
+    u = reinterpret(UInt64, t)
+    u = (u + 0x8000_0000) & UInt64(0xffff_ffff_c000_0000)
+    t = reinterpret(Float64, u)
+
+    # one step Newton iteration solving
+    #   t^3 - x == 0
+    # with update
+    #   t <- t + t * (x/t^2 - t) / (3*t)
+
+    # to 53 bits with error < 0.667 ulps
+    s = t*t             # t*t is exact
+    r = x/s             # error <= 0.5 ulps; |r| < |t|
+    w = t+t             # t+t is exact
+    r = (r - t)/(w + r) # r-t is exact; w+r ~= 3*t
+    t = muladd(t, r, t) # error <= 0.5 + 0.5/3 + epsilon
+    return t
+end
+
+function cbrt(x::Union{Float32,Float64})
+    if !isfinite(x) || x == zero(x)
```

Comment:
`iszero(x)`, though it probably makes no difference

Reply:
i forgot we had `iszero`

Reply:
Writing `x + x === x` might be a tricksy way of writing this condition. Analysis:
* `-0.0 + -0.0 === -0.0`
* `0.0 + 0.0 === 0.0`
* `Inf + Inf === Inf`
* `-Inf + -Inf === -Inf`
* `NaN + NaN === NaN`

It should be false for any non-zero finite value. The last one is subtle one: the result of `+` when both arguments are NaN will be one of the arguments which are both bitwise identical to each other and thus to the result. I forget if this is guaranteed by the IEEE standard or not; I also forget if `===` requires moving to an integer register or if there's an fp operation for exact bit comparison. If not, then we could just do `isequal(x + x, x)` instead.